src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -270,6 +345,32 @@ internal static bool TryGetMinimalCSharpRange(RazorCodeDocument codeDocument, Ra
         return false;
     }
 
+    internal static bool TryGetCSharpRanges(RazorCodeDocument codeDocument, Range razorRange, out Range[] ranges)
+    {
+        using var _ = ListPool<Range>.GetPooledObject(out var csharpRanges);
+        var csharpSourceText = codeDocument.GetCSharpSourceText();
+        var sourceText = codeDocument.GetSourceText();
+        var textSpan = razorRange.AsTextSpan(sourceText);
+        var csharpDoc = codeDocument.GetCSharpDocument();
+        csharpRanges.SetCapacityIfLarger(csharpDoc.SourceMappings.Count());
```

Comment:
>Count() [](http://example.com/codeflow?start=66&length=7)

Big red flag when calling the Count extension method. I'd just remove this line and let the items in the cache resize if they need to.

Reply:
particularly because the number of source mappings could greatly exceed the number of interlapped source mappings.

Reply:
Yep, agreed. This is probably my fault for suggestion that we should set the capacity, but I forgot `SourceMappings` is not a concrete type. Good news is we can fix that now that 17.8 P1 is on our build images, so we can follow up and add this back later.

Reply:
I'm learning that concrete types allow all kinds of goodness. Still here, I might avoid doing the capacity setting because that number could be *significantly* larger than the number of intersecting regions.

Reply:
I supppose for large Razor files you're right, there could be a lot out of range.

We can always leave this for now, and put something back if it shows up in a trace.